library/bignum.c

```diff
@@ -2421,12 +2444,92 @@ int mbedtls_mpi_fill_random( mbedtls_mpi *X, size_t size,
 
-    Xp = (unsigned char*) X->p;
-    MBEDTLS_MPI_CHK( f_rng( p_rng, Xp + overhead, size ) );
+    ret = mpi_fill_random_internal( X, size, f_rng, p_rng );
 
-    mpi_bigendian_to_host( X->p, limbs );
+cleanup:
+    return( ret );
+}
+
+int mbedtls_mpi_random( mbedtls_mpi *X,
+                        mbedtls_mpi_sint min,
+                        const mbedtls_mpi *N,
+                        int (*f_rng)(void *, unsigned char *, size_t),
+                        void *p_rng )
+{
+    int ret = MBEDTLS_ERR_MPI_BAD_INPUT_DATA;
+    int count;
+    unsigned cmp = 0;
+    size_t n_bits = mbedtls_mpi_bitlen( N );
+    size_t n_bytes = ( n_bits + 7 ) / 8;
+
+    if( min < 0 )
+        return( MBEDTLS_ERR_MPI_BAD_INPUT_DATA );
+    if( mbedtls_mpi_cmp_int( N, min ) <= 0 )
+        return( MBEDTLS_ERR_MPI_BAD_INPUT_DATA );
+
+    /*
+     * When min == 0, each try has at worst a probability 1/2 of failing
+     * (the msb has a probability 1/2 of being 0, and then the result will
+     * be < N), so after 30 tries failure probability is a most 2**(-30).
+     *
+     * When N is just below a power of 2, as is the case when generating
+     * a random scalar on most elliptic curves, 1 try is enough with
+     * overwhelming probability. When N is just above a power of 2,
+     * as when generating a random scalar on secp224k1, each try has
+     * a probability of failing that is almost 1/2.
+     *
+     * The probabilities are almost the same if min is nonzero but negligible
+     * compared to N. This is always the case when N is crypto-sized, but
+     * it's convenient to support small N for testing purposes. When N
+     * is small, use a higher repeat count, otherwise the probability of
+     * failure is macroscopic.
+     */
+    if( n_bytes <= 4 )
+        count = 250;
+    else
+        count = 30;
+
+    /* Ensure that target MPI has exactly the same number of limbs
+     * as the upper bound, even if it has leading zeros.
+     * This is necessary for the mbedtls_mpi_lt_mpi_ct() check. */
+    if( X->n != N->n )
+    {
+        mbedtls_mpi_free( X );
+        mbedtls_mpi_init( X );
+        MBEDTLS_MPI_CHK( mbedtls_mpi_grow( X, N->n ) );
+    }
+    X->s = 1;
+
+    /*
+     * Match the procedure given in RFC 6979 §3.3 (deterministic ECDSA)
+     * when f_rng is a suitably parametrized instance of HMAC_DRBG:
+     * - use the same byte ordering;
+     * - keep the leftmost n_bits bits of the generated octet string;
+     * - try until result is in the desired range.
+     * This also avoids any bias, which is especially important for ECDSA.
+     */
+    do
+    {
+        MBEDTLS_MPI_CHK( mpi_fill_random_internal( X, n_bytes, f_rng, p_rng ) );
+        MBEDTLS_MPI_CHK( mbedtls_mpi_shift_r( X, 8 * n_bytes - n_bits ) );
+
+        if( --count == 0 )
+        {
+            ret = MBEDTLS_ERR_MPI_NOT_ACCEPTABLE;
+            goto cleanup;
+        }
+
+        ret = mbedtls_mpi_lt_mpi_ct( X, N, &cmp );
+        if( ret != 0 )
+        {
```

Comment:
Knowing the internal behavior of `mbedtls_mpi_lt_mpi_ct`: yes, it can't fail here. But based only on the documented behavior, I don't see how the check could be omitted: if the function fails, the contents of `cmp` is unspecified, so we don't know whether the loop will continue.

Reply:
I was just suggesting to remove the unnecessary curly brackets thus:
```c
if( ret != 0 )
    goto cleanup;
```

Reply:
Ah ok. That's cosmetic, so I'm not going to bother changing it unless I change this part of the code for another reason.

Reply:
And by the way probably even better: MBEDTLS_MPI_CHK( mbedtls_mpi_lt_mpi_ct( ...

Reply:
Actually the change to `MBEDTLS_MPI_CHK()` is made in https://github.com/ARMmbed/mbedtls-restricted/pull/820 (last commit currently) so perhaps we can skip it for this PR?

Reply:
Yes, `MBEDTLS_MPI_CHK` does make sense. Changing.